File: src/client/framework/pacer.ts
/**
 * Fixed-step accumulator for labs WITHOUT a reconciler (`predict.tick()` only
 * paces once a reconciler adopts the fixed step; a prediction-free client
 * still has to send one input per server tick).
 */
export class FixedStepPacer {
  private acc = 0;
  private last = -1;

  constructor(private stepMs: number) {}

  steps(now: number): number {
    if (this.last < 0) { this.last = now; return 0; }
    this.acc += now - this.last;
    this.last = now;
    let n = Math.floor(this.acc / this.stepMs);
    if (n > 5) { n = 5; this.acc = 0; }   // hitch (tab switch): drop the backlog
    else this.acc -= n * this.stepMs;
    return n;
  }

  reset(): void { this.acc = 0; this.last = -1; }
}

File: clients/flutter-app/lib/pacer.dart
/// Fixed-step accumulator for labs WITHOUT a reconciler: `predict.tick()` only
/// paces once a reconciler adopts the fixed step, but a prediction-free client
/// still has to send one input per server tick.
///
/// Port of `godot-gd-app/scripts/pacer.gd`.
class FixedStepPacer {
  FixedStepPacer(this.stepMs);

  /// Milliseconds per fixed step (`1000 / TICK_HZ` for the playground).
  final double stepMs;

  double _acc = 0.0;
  double _last = 0.0;
  bool _started = false;

  /// Forget the accumulator — after a reconnect, or a deliberate stall.
  void reset() {
    _acc = 0.0;
    _started = false;
  }

  /// Steps due at [nowMs]. The first call only latches the clock (0 steps).
  int steps(double nowMs) {
    if (!_started) {
      _started = true;
      _last = nowMs;
      return 0;
    }
    _acc += nowMs - _last;
    _last = nowMs;
    var n = (_acc / stepMs).floor();
    if (n > 5) {
      // Hitch (a stall, a breakpoint): drop the backlog rather than firing a
      // burst of inputs the server would reject as a flood.
      n = 5;
      _acc = 0.0;
    } else {
      _acc -= n * stepMs;
    }
    return n;
  }
}
